### PR TITLE
plumbing: protocol/packp, client-side filter capability support

### PR DESCRIPTION
**Pull request by edigaryev**
This allows lower-level `go-git` package consumers to interact with registries supporting `filter` capability and partially fetch objects given a `Filter` specification.

Related: https://github.com/go-git/go-git/issues/713.

### CODE REVIEW

File: plumbing/protocol/packp/filter.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	return Filter(fmt.Sprintf("blob:limit=%d", n))
+}
+
+// FilterTreeDepth omits all blobs and trees whose depth from the root tree
+// is larger or equal to depth.
+func FilterTreeDepth(depth uint64) Filter {
+	return Filter(fmt.Sprintf("tree:%d", depth))
+}
+
+// FilterObjectType omits all objects which are not of the requested type t.
```
Comment (pjbgf) [inline plumbing/protocol/packp/filter.go:31]:
```suggestion
// FilterObjectType omits all objects which are not of the requested type t.
// Supported types are TagObject, CommitObject, TreeObject and BlobObject.
```

File: plumbing/protocol/packp/filter.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+func FilterObjectType(t plumbing.ObjectType) Filter {
```
Comment (pjbgf) [inline plumbing/protocol/packp/filter.go:32]: We need to add validation here and return `error` if an unsupported ObjectType (InvalidObject, OFSDeltaObject, REFDeltaObject or AnyObject) is used.

**Review by pjbgf — COMMENTED** [review]
@edigaryev great job on putting this together. Apart from nits below it LGTM.

File: plumbing/protocol/packp/filter.go (hunk 2 above)
Comment (edigaryev) [reply plumbing/protocol/packp/filter.go]: Implemented in https://github.com/go-git/go-git/pull/1000/commits/83f87a265c4dd01dbeedfaf99517fa000b4dbd98.

File: plumbing/protocol/packp/filter.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+// FilterBlobNone omits all blobs.
+func FilterBlobNone() Filter {
+	return "blob:none"
+}
+
+// FilterBlobLimit omits blobs of size at least n bytes. n can be zero,
+// in which case all blobs will be omitted.
+func FilterBlobLimit(n uint64) Filter {
+	return Filter(fmt.Sprintf("blob:limit=%d", n))
```
[the remark this answers is not in the document]
Comment (edigaryev) [reply plumbing/protocol/packp/filter.go]: Implemented in https://github.com/go-git/go-git/pull/1000/commits/84d62395c338d4dd89ad3f9946fbda798ee33a73.

File: plumbing/protocol/packp/filter.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import (
+	"fmt"
+	"github.com/go-git/go-git/v5/plumbing"
+	"net/url"
+	"strings"
+)
+
+// Filter values enable the partial clone capability which causes
+// the server to omit objects that match the filter.
+type Filter string
```
[the remark this answers is not in the document]
Comment (edigaryev) [reply plumbing/protocol/packp/filter.go:19]: Added in https://github.com/go-git/go-git/pull/1000/commits/147c8b096a4469a2102d097e466f7f1f2d73d226.

**Review by 0xDEC0DE — APPROVED** [review]

**Comment by edigaryev** [comment]
@pjbgf would you mind taking a look at this once more?

**Comment by pjbgf** [comment]
For future reference, this does not fully implement the [partial clone spec](https://github.com/git/git/blob/master/Documentation/technical/partial-clone.txt), and follow-up PRs will be needed until this is exposed via `FetchOptions`/`CloneOptions`.

**Review by pjbgf — APPROVED** [review]
@edigaryev great stuff, thanks for working on this. :1st_place_medal:

**Pull request merged**